### PR TITLE
feat: validate manifest before writing

### PR DESCRIPTION
**Pull request by letFunny**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Review by rebornplusplus — COMMENTED** [review]
Looks good to me! I left a comment below about a function name.

File: internal/manifest/manifest.go [hunk 1]
```diff
@@ -301,3 +306,104 @@ func unixPerm(mode fs.FileMode) (perm uint32) {
 	}
 	return perm
 }
+
+func fastValidate(options *WriteOptions) (err error) {
```
[the remark this answers is not in the document]
Comment (rebornplusplus) [reply internal/manifest/manifest.go:310]: A comment would be nice here as well. :)

[the remark this answers is not in the document]
Comment (letFunny) [reply internal/manifest/manifest.go:311]: Agree, comment added.

File: internal/manifest/manifest.go [hunk 2]
```diff
@@ -301,3 +306,105 @@ func unixPerm(mode fs.FileMode) (perm uint32) {
 	}
 	return perm
 }
+
+// fastvalidate validates the data to be written into the manifest on disk.
```
Comment (niemeyer) [inline internal/manifest/manifest.go:310]: Can we please add:
```
// fastValidate validates the data to be written into the manifest.
// This is validating internal structures which are supposed to be correct unless there is
// a bug. As such, only assertions that can be done quickly are performed here, instead
// of it being a comprehensive validation of all the structures.
```

File: internal/manifest/manifest_test.go [hunk 3]
```diff
@@ -265,16 +266,26 @@ func (s *S) TestFindPaths(c *C) {
...
+var slice1 = &setup.Slice{
+	Package: "package1",
+	Name:    "slice1",
+}
+var slice2 = &setup.Slice{
+	Package: "package2",
+	Name:    "slice2",
+}
+
+var generateManifestTests = []struct {
```
Comment (niemeyer) [inline internal/manifest/manifest_test.go:278]: Thanks for the new tests.

**Review by niemeyer — APPROVED** [review]
Very nice, thanks. Only a trivial adding some notes to fastValidate.

File: internal/manifest/manifest.go (hunk 2 above)
Comment (niemeyer) [reply internal/manifest/manifest.go]: I went ahead and edited it as it's just an additional comment.

**Comment by endersonmaia** [comment]
Sorry to make the question here.

What's the plan, add a an essential manifest slice to all packages like we have for copyright?

ex.:

```yaml
package: libc6

essential:
  - libc6_copyright
  - libc6_manifest

slices:
(...)

  copyright:
    contents:
      /usr/share/doc/libc6/copyright:

  manifest:
    contents:
      /var/lib/chisel/**: {generate:  manifest}
```

**Comment by rebornplusplus** [comment]
> `base-files_chisel` is a slice that when installed will feed the manifest with information from all other package slices installed together in the same `chisel cut` action.

Correct.

> And `*_manifest` is a "virtual" slice that's available for all packages, that when installed will feed the manifest with information from all other slices that are its dependencies on the same `chisel cut` action.

Nope. `*_manifest` does not exist, not even virtually. You may create those, similar to `base-files_chisel`. But you don't really need multiple slices like `base-files_chisel`, unless you want to create multiple manifest file with (almost, if not) the same contents.

**Pull request merged**